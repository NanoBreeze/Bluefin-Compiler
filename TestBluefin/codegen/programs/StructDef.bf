/*
Verifies struct definitions, including nested members
Verifies varDecl with expr() inside struct (fields)
TODO: Add struct member access, methods, inheritance, and polymorphism
*/

int blow = 8;
int here = blow + 10;

int nice() {
	return 1;
}

struct Hey {
	int a;
	int b = a;
	float c = 123 * a;
};


struct Hi {
	int a;
	//int aa = a;
	Hey hey;
	//Hey hi = hey;
	bool b = a > blow;
};

void foo() {
	Hey hey;  // in order for StructDefs to show up, we must make a local variable with them
	Hi hi;
}