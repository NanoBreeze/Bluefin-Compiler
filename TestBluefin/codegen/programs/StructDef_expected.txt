; ModuleID = 'StructDef.bf'
source_filename = "StructDef.bf"

%Normal = type { i32, float, i32, i32, float }
%Forwards = type { i32, float, i32, i32 }
%MethodCalls = type {}
%StructMemberDecl = type { %Normal, float, i32 }

@globalVar = global i32 8
@norm = external global %Normal
@llvm.global_ctors = appending global [1 x { i32, void ()*, i8* }] [{ i32, void ()*, i8* } { i32 0, void ()* @_VarDeclInitializations, i8* null }]

define i32 @globalFunc() {
entry:
  ret i32 1
}

define i32 @_Normal_method(%Normal* %this, i32 %a) {
entry:
  %alloctmp_a = alloca i32, align 4
  store i32 %a, i32* %alloctmp_a, align 4
  %loadtmp_a = load i32, i32* %alloctmp_a, align 4
  %casttmp = sitofp i32 %loadtmp_a to float
  %memPtr_b = getelementptr %Normal, %Normal* %this, i32 0, i32 1
  %loadtmp_b = load float, float* %memPtr_b, align 4
  %addtmp = fadd float %casttmp, %loadtmp_b
  %loadtmp_globalVar = load i32, i32* @globalVar, align 4
  %casttmp1 = sitofp i32 %loadtmp_globalVar to float
  %addtmp2 = fadd float %addtmp, %casttmp1
  ret i32 2
}

define void @_ctor_Normal(%Normal* %this) {
entry:
  %memPtr_a = getelementptr %Normal, %Normal* %this, i32 0, i32 0
  %loadtmp_a = load i32, i32* %memPtr_a, align 4
  %addtmp = add i32 %loadtmp_a, 6
  %casttmp = sitofp i32 %addtmp to float
  %memPtr_b = getelementptr %Normal, %Normal* %this, i32 0, i32 1
  store float %casttmp, float* %memPtr_b, align 4
  %calltmp = call i32 @_Normal_method(%Normal* %this, i32 99)
  %memPtr_orange = getelementptr %Normal, %Normal* %this, i32 0, i32 2
  store i32 %calltmp, i32* %memPtr_orange, align 4
  %calltmp1 = call i32 @globalFunc()
  %memPtr_c = getelementptr %Normal, %Normal* %this, i32 0, i32 3
  store i32 %calltmp1, i32* %memPtr_c, align 4
  %loadtmp_globalVar = load i32, i32* @globalVar, align 4
  %memPtr_a2 = getelementptr %Normal, %Normal* %this, i32 0, i32 0
  %loadtmp_a3 = load i32, i32* %memPtr_a2, align 4
  %multmp = mul i32 %loadtmp_globalVar, %loadtmp_a3
  %calltmp4 = call i32 @_Normal_method(%Normal* %this, i32 1)
  %addtmp5 = add i32 %multmp, %calltmp4
  %casttmp6 = sitofp i32 %addtmp5 to float
  %memPtr_d = getelementptr %Normal, %Normal* %this, i32 0, i32 4
  store float %casttmp6, float* %memPtr_d, align 4
  ret void
}

define void @_Forwards_there(%Forwards* %this) {
entry:
  %memPtr_b = getelementptr %Forwards, %Forwards* %this, i32 0, i32 1
  %loadtmp_b = load float, float* %memPtr_b, align 4
  %memPtr_d = getelementptr %Forwards, %Forwards* %this, i32 0, i32 3
  %loadtmp_d = load i32, i32* %memPtr_d, align 4
  %casttmp = sitofp i32 %loadtmp_d to float
  %subtmp = fsub float %loadtmp_b, %casttmp
  ret void
}

define i32 @_Forwards_first(%Forwards* %this) {
entry:
  ret i32 1
}

define i32 @_Forwards_second(%Forwards* %this, i32 %a, i32 %c) {
entry:
  %alloctmp_a = alloca i32, align 4
  store i32 %a, i32* %alloctmp_a, align 4
  %alloctmp_c = alloca i32, align 4
  store i32 %c, i32* %alloctmp_c, align 4
  %loadtmp_a = load i32, i32* %alloctmp_a, align 4
  %loadtmp_c = load i32, i32* %alloctmp_c, align 4
  %addtmp = add i32 %loadtmp_a, %loadtmp_c
  %memPtr_d = getelementptr %Forwards, %Forwards* %this, i32 0, i32 3
  %loadtmp_d = load i32, i32* %memPtr_d, align 4
  %addtmp1 = add i32 %addtmp, %loadtmp_d
  %loadtmp_globalVar = load i32, i32* @globalVar, align 4
  %addtmp2 = add i32 %addtmp1, %loadtmp_globalVar
  ret i32 %addtmp2
}

define void @_ctor_Forwards(%Forwards* %this) {
entry:
  %calltmp = call i32 @_Forwards_first(%Forwards* %this)
  %memPtr_a = getelementptr %Forwards, %Forwards* %this, i32 0, i32 0
  store i32 %calltmp, i32* %memPtr_a, align 4
  %memPtr_a1 = getelementptr %Forwards, %Forwards* %this, i32 0, i32 0
  %loadtmp_a = load i32, i32* %memPtr_a1, align 4
  %memPtr_c = getelementptr %Forwards, %Forwards* %this, i32 0, i32 2
  %loadtmp_c = load i32, i32* %memPtr_c, align 4
  %calltmp2 = call i32 @_Forwards_second(%Forwards* %this, i32 %loadtmp_a, i32 %loadtmp_c)
  %casttmp = sitofp i32 %calltmp2 to float
  %memPtr_b = getelementptr %Forwards, %Forwards* %this, i32 0, i32 1
  store float %casttmp, float* %memPtr_b, align 4
  %memPtr_d = getelementptr %Forwards, %Forwards* %this, i32 0, i32 3
  %loadtmp_d = load i32, i32* %memPtr_d, align 4
  %memPtr_c3 = getelementptr %Forwards, %Forwards* %this, i32 0, i32 2
  store i32 %loadtmp_d, i32* %memPtr_c3, align 4
  ret void
}

define void @_MethodCalls_first(%MethodCalls* %this) {
entry:
  %alloctmp_a = alloca i32, align 4
  %calltmp = call i32 @_MethodCalls_third(%MethodCalls* %this, i32 2)
  store i32 %calltmp, i32* %alloctmp_a, align 4
  %loadtmp_a = load i32, i32* %alloctmp_a, align 4
  %calltmp1 = call i32 @_MethodCalls_third(%MethodCalls* %this, i32 3)
  store i32 %calltmp1, i32* %alloctmp_a, align 4
  ret void
}

define void @_MethodCalls_second(%MethodCalls* %this) {
entry:
  call void @_MethodCalls_first(%MethodCalls* %this)
  %calltmp = call i32 @_MethodCalls_third(%MethodCalls* %this, i32 1)
  %calltmp1 = call i32 @_MethodCalls_third(%MethodCalls* %this, i32 %calltmp)
  %calltmp2 = call i32 @_MethodCalls_third(%MethodCalls* %this, i32 %calltmp1)
  ret void
}

define i32 @_MethodCalls_third(%MethodCalls* %this, i32 %a) {
entry:
  %alloctmp_a = alloca i32, align 4
  store i32 %a, i32* %alloctmp_a, align 4
  ret i32 1
}

define void @_ctor_MethodCalls(%MethodCalls* %this) {
entry:
  ret void
}

define void @_ctor_StructMemberDecl(%StructMemberDecl* %this) {
entry:
  %memPtr_norm = getelementptr %StructMemberDecl, %StructMemberDecl* %this, i32 0, i32 0
  call void @_ctor_Normal(%Normal* %memPtr_norm)
  %memPtr_norm1 = getelementptr %StructMemberDecl, %StructMemberDecl* %this, i32 0, i32 0
  %loadtmp_norm = load %Normal, %Normal* %memPtr_norm1, align 4
  %memPtr_d = getelementptr %Normal, %Normal* %memPtr_norm1, i32 0, i32 4
  %loadtmp_d = load float, float* %memPtr_d, align 4
  %memPtr_d2 = getelementptr %StructMemberDecl, %StructMemberDecl* %this, i32 0, i32 1
  store float %loadtmp_d, float* %memPtr_d2, align 4
  %memPtr_norm3 = getelementptr %StructMemberDecl, %StructMemberDecl* %this, i32 0, i32 0
  %loadtmp_norm4 = load %Normal, %Normal* %memPtr_norm3, align 4
  %calltmp = call i32 @_Normal_method(%Normal* %memPtr_norm3, i32 2)
  %memPtr_a = getelementptr %StructMemberDecl, %StructMemberDecl* %this, i32 0, i32 2
  store i32 %calltmp, i32* %memPtr_a, align 4
  ret void
}

define internal void @internalGlobalVarDecl_norm() {
entry:
  call void @_ctor_Normal(%Normal* @norm)
  ret void
}

define void @foo() {
entry:
  %alloctmp_normal = alloca %Normal, align 8
  call void @_ctor_Normal(%Normal* %alloctmp_normal)
  %alloctmp_forwards = alloca %Forwards, align 8
  call void @_ctor_Forwards(%Forwards* %alloctmp_forwards)
  %alloctmp_methodCalls = alloca %MethodCalls, align 8
  call void @_ctor_MethodCalls(%MethodCalls* %alloctmp_methodCalls)
  %alloctmp_decl = alloca %StructMemberDecl, align 8
  call void @_ctor_StructMemberDecl(%StructMemberDecl* %alloctmp_decl)
  ret void
}

define internal void @_VarDeclInitializations() {
entry:
  call void @internalGlobalVarDecl_norm()
  ret void
}
